testutil/ext_err.go

```diff
@@ -0,0 +1,74 @@
+package testutil
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"runtime"
+	"strings"
+)
+
+const envFlag = "GH_ANNOTATION"
+
+// ExtError is a wrapper to the error type, extending it with the file and line
+// that produced the error, and a generic error description that is sent to stdout
+// as a Github annotation when the envFlag environment variable is set.
+type ExtError struct {
+	err  error
+	file string
+	line int
+	desc string
+}
+
+func newExtError(err error, desc string) ExtError {
+	_, fn, ln, ok := runtime.Caller(2)
+	if !ok {
+		panic("Couldn't recover runtime info")
+	}
+	fileName := fn[strings.LastIndex(fn, "/linkerd2/")+10:]
+	return ExtError{
+		err:  err,
+		file: fileName,
+		line: ln,
+		desc: desc,
+	}
+}
+
+// Error is a wrapper for test error messages; msg can either be
+// a string or an error type.
+func Error(msg interface{}) ExtError {
+	switch v := msg.(type) {
+	case error:
+		return newExtError(v, v.Error())
+	case string:
+		return newExtError(errors.New(v), v)
+	default:
+		panic("Invalid type calling testutil.Error()")
+	}
+
+}
+
+// Errorf is a wrapper for test error messages, following
+// the Printf() signature (using a format specifier)
+func Errorf(format string, args ...interface{}) ExtError {
+	err := fmt.Errorf(format, args...)
+	return newExtError(err, err.Error())
+}
+
+// WithAnn is to be called on a ExtError type to provide a generic
+// error description
+func (extErr ExtError) WithAnn(desc string) ExtError {
+	extErr.desc = desc
+	return extErr
+}
+
+// Error ensures that the struct ExtError implements the `error` interface.
```

Comment:
Things will be a bit weird if callers do anything with this Error other than just immediately pass it to `t.Error`.  Anyone calling `.Error()` on it will cause an annotation to be printed at that time.

Reply:
I could have it check in the call stack whether the call comes from `testing.Error`, and only in that case produce the annotation.
That being said, I share the feeling that all this is a bit hacky, but I can't figure an alternative way while conserving the default testing framework for go...

Reply:
Ok this should no longer be a concern if I adopt the `testutil.Fatalf()` approach you suggest above, dumping the `ExtError` struct :+1: